### PR TITLE
Multipart pipe reader

### PR DESCRIPTION
**Pull request by gilmishal**
Continuing work on MultipartForm PipeReader Implementation from PR #15056.

Addresses #11377

### CODE REVIEW

File: src/Http/WebUtilities/test/MultipartPipeReaderTests.cs [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+            Assert.Equal("form-data; name=\"file2\"; filename=\"a.html\"", section!.Headers!["Content-Disposition"][0]);
+            Assert.Equal("text/html", section!.Headers!["Content-Type"][0]);
+            result = await section!.ReadAsStringAsync();
+            Assert.Equal("<!DOCTYPE html><title>Content of a.html.</title>\r\n", result);
+
+            Assert.Null(await reader.ReadNextSectionAsync());
+        }
+
+        [Fact]
+        public async Task MultipartPipeReader_TwoPartBodyIncompleteBuffer_OneSectionReadsSuccessfullySecondSectionThrows()
```
Comment (gilmishal) [inline src/Http/WebUtilities/test/MultipartPipeReaderTests.cs:272]: I fixed the test to pass, but I am not sure if it's a functional requirement or not - Adding a subsequent AdvanceTo and ReadAsync will throw, but I am not sure if that's what we want. 
Basically, it's conflicting with a comment inside the MultipartSectionPipeReader  _We want them to consume and advance all of the data before we try to drain the metadata / boundary._

File: src/Http/WebUtilities/src/MultipartSectionPipeReader.cs [hunk 2]
```diff
@@ -73,11 +73,21 @@ public override void AdvanceTo(SequencePosition consumed, SequencePosition exami
...
+                {
+                    // We may have a partial boundary match in the
+                    // unconsumed buffer and we need more data to continue.
+                    _pipeReader.AdvanceTo(consumed, _unconsumedData.End);
+                }
+                else
+                {
+                    // We reached the end of the section, we should advance to the end.
+                    _pipeReader.AdvanceTo(_unconsumedData.End);
+                    _finalLineConsumed = true;
```
Comment (gilmishal) [inline src/Http/WebUtilities/src/MultipartSectionPipeReader.cs:89]: This solves an issue with _finalLineConsumed is false (and probably shouldn't be), I am not sure if it's the correct solution though, and I might be missing an edge case.

File: src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            Assert.True(result.IsCompleted);
+            Assert.False(result.IsCanceled);
+            Assert.True(result.Buffer.IsEmpty);
+        }
+
+        [Theory]
+        [InlineData(TextAndBoundary, Text, 1)]
+        [InlineData(HtmlWithNewLinesAndBoundary, HtmlWithNewLines, 52)]
+        [InlineData(TextWithPartialBoundaryMatchAndBoundary, TextWithPartialBoundaryMatch, 14)]
+        public async Task MultipartSectionPipeReader_SingleByteReadWithoutAdvance_Success(string input, string expected, int reads)
```
Comment (gilmishal) [inline src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs:107]: Not sure how it differs from  WithAdvance Test. The only difference is that the result isn't buffered and the expected bytes read is larger - both seem like a mistake in the test rather than the functionality, since I am not sure how it's even supposed to pass - **I am gonna guess that this is a copy pasted test that was never implemented, but I am double checking.**

File: src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        [InlineData(TextAndBoundary, Text)]
+        [InlineData(HtmlWithNewLinesAndBoundary, HtmlWithNewLines)]
+        [InlineData(TextWithPartialBoundaryMatchAndBoundary, TextWithPartialBoundaryMatch)]
+        public async Task MultipartSectionPipeReader_ValidBody_Success(string input, string expected)
+        {
+            var pipeReader = MakeReader(input);
+            var sectionReader = new MultipartSectionPipeReader(pipeReader, new MultipartBoundary(Boundary));
+
+            var result = await sectionReader.ReadAsync();
+            Assert.False(result.IsCompleted);
```
Comment (gilmishal) [inline src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs:58]: IsCompleted currently comes out as true, This is because it buffered the entire section on the first Read to the underlying Pipe, and it truly reached the end of the section. I am not sure if that's a problem or not.

**Comment by gilmishal** [comment]
@Tratcher I have made most  of the unit tests pass - there are some tests that are still failing, but I need your help deciding what to do with these tests. 

I feel that the Pipe Implementation itself is solid (not sure about the test coverage though), but we still need to decide how to buffer (and deliver) the sections, to allow reading the sections interchangeably.

**Comment by En3Tho** [comment]
Can you post benchmark results here please? Just to see the effect in plain numbers

File: src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs (hunk 3 above)
Comment (Tratcher) [reply src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs:107]: The difference is that the prior test always calls `AdvanceTo(temp.Buffer.End);` but this one calls `AdvanceTo(temp.Buffer.Start, temp.Buffer.End);`. That makes the former unbuffered and the latter buffered. Both are valid scenarios.

File: src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs (hunk 4 above)
Comment (Tratcher) [reply src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs:58]: It's fine for IsComplete to return true so long as all of the data is available and we've fully parsed the boundary. Just make sure AdvanceTo(end) also handles advancing the underlying pipe past the parsed boundary.

**Review by Tratcher — COMMENTED** [review]
I've only skimmed the code and answered questions so far.

**Comment by gilmishal** [comment]
> Can you post benchmark results here please? Just to see the effect in plain numbers

I am unable to run any of the benchmarking projects - I am getting this error 
`Project Microsoft.AspNetCore.WebUtilities.Performance is not compatible with net5.0`

I am not sure what's causing it, and I couldn't find any relevant documentation in this repo, or on BenchmarkDotNet.

I can say that I did run activate.ps1 assuming it was related to the cli dotnet version being used (apperantly not though) - but I do think it might be related to this lines in Program.cs (in the performance projects)

 ```
.WithToolchain(CsProjCoreToolchain.From(new NetCoreAppSettings
                    (
                        // not using "net6.0", a workaround for https://github.com/dotnet/BenchmarkDotNet/pull/1479
                        targetFrameworkMoniker: "netcoreapp5.0",
                        runtimeFrameworkVersion: default,
                        name: ".NET Core 5.0"
                    )))
```

I don't have any other angle at the moment, other than perhaps trying to build this thing for .net 5.

File: src/Http/WebUtilities/test/MultipartPipeReaderTests.cs (hunk 1 above)
Comment (gilmishal) [reply src/Http/WebUtilities/test/MultipartPipeReaderTests.cs:272]: Looking at the original now, I am not sure I would say that the behavior had changed.
If we alter the original test and add another read to the section.Body it would fail.
If we alter this test to read the third section instead of continuing to read the rest of the second section it would also fail.

I think that this behavior was changed with my original implementation of the PipeReader, and the test just wasn't updated.

Anyways, we might want to consider altering the test to reflect the actual scenario - basically the behavior in both the Stream implementation and the Pipe Implementation is that for an incomplete Section Boundry - the second section will throw before reaching the end - (Pipe - IsCompleted == true, Stream - Read returns 0) - and trying to read the third section will throw.  I would argue that in reality we are more likely to try and read the second section to it's end before attempting to read the third section, especially since we will probably copy each section to it's own Pipe.

File: src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs (hunk 3 above)
Comment (gilmishal) [reply src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs:107]: Oh, I might have missed that. Anyways, I was wondering why they were expecting different number of reads, since they are both pretty similar. I am guessing that this was a mistake.

File: src/Http/WebUtilities/src/MultipartPipeSection.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        /// <summary>
+        /// Gets or sets the body.
+        /// </summary>
+        public PipeReader BodyReader { get; set; } = default!;
+
+        /// <summary>
+        /// The position where the body starts in the total multipart body.
+        /// This may not be available if the total multipart body is not seekable.
+        /// </summary>
+        public long? BaseStreamOffset { get; set; }
```
[the remark this answers is not in the document]
Comment (gilmishal) [reply src/Http/WebUtilities/src/MultipartPipeSection.cs:57]: removed

File: src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs (hunk 4 above)
Comment (gilmishal) [reply src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs]: Ok then in that case, I think the test is fixed and covers this case.

**Comment by gilmishal** [comment]
I have addressed your comments and fixed the tests accordingly, please verify that you approve my changes, and with your permission I will fix TryRead and create the tests for it.
Then, I will start working on buffering the sections.

File: src/Http/WebUtilities/src/MultipartSectionPipeReader.cs [hunk 6]
```diff
@@ -180,7 +180,7 @@ public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancella
                 if (!_boundaryFound)
                 {
                     // We must be examining a partial boundary match, keep reading.
-                    _pipeReader.AdvanceTo(_unconsumedData.Start, _unconsumedData.End);
+                    _pipeReader.AdvanceTo(readResult.Buffer.Start, _unconsumedData.End);
```
Comment (gilmishal) [inline src/Http/WebUtilities/src/MultipartSectionPipeReader.cs:183]: This was originally _bodyBuffer.Start - I changed it to _unconsumedData.Start, which is totally wrong because it advanced the pipe without the user advancing it.

The reason I changed it was that _bodyBuffer.Start was resulting in a never ending loop - that happend because if the read isn't long enough to check if it's a boundary - it wouldn't set the _bodyBuffer (which does make sense)  - since _bodyBuffer was unset, the advanceTo wouldn't actually advance, and create a never ending loop.

I think that readResult.Buffer.Start would avoid this issue, and reflects the actual meaning (that's why I changed all advances to use it instead of _bodyBuffer.Start - Just informing you so you find it easier to understand the changes I made.

**Comment by davidfowl** [comment]
👀

File: src/Http/WebUtilities/src/MultipartPipeSection.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+
```
Comment (Tratcher) [inline src/Http/WebUtilities/src/MultipartPipeSection.cs:19]:
```suggestion
```

File: src/Http/WebUtilities/perf/Microsoft.AspNetCore.WebUtilities.Performance/MultipartReaderBenchmark.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using BenchmarkDotNet.Attributes;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    public class MultipartReaderBenchmark
+    {
+
```
Comment (Tratcher) [inline src/Http/WebUtilities/perf/Microsoft.AspNetCore.WebUtilities.Performance/MultipartReaderBenchmark.cs:18]:
```suggestion
```

**Review by Tratcher — COMMENTED** [review]
Here's a little more detailed feedback. I didn't get to the most complex part in MultipartSectionPipeReader.

I talked this over with our team and we reached two conclusions:
1) We're starting our 6.0 development and focusing on P0 features for the first milestone or two. We do think this multipart pipe reader is important for 6.0, but it will be a milestone or two before we're able to focus on it.
2) @jkotalik has a lot more experience with custom pipe readers so he'll take over reviewing this PR.

File: src/Http/WebUtilities/src/MultipartPipeReader.cs [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+                }
+
+                var buffer = readResult.Buffer;
+                try
+                {
+                    var finishedParsing = TryParseHeadersToEnd(ref buffer, ref headersAccumulator, ref headersLength);
+                    if (headersLength > HeadersLengthLimit)
+                    {
+                        throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");
+                    }
```
[the remark this answers is not in the document]
Comment (gilmishal) [reply src/Http/WebUtilities/src/MultipartPipeReader.cs:110]: yes, it's redundant I'll remove.

File: src/Http/WebUtilities/src/MultipartPipeReader.cs [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+
+                    if (readResult.IsCompleted)
+                    {
+                        throw new InvalidDataException("Unexpected end of Stream, the content may have already been read by another component. ");
+                    }
+                }
+                finally
+                {
+                    _pipeReader.AdvanceTo(buffer.Start, buffer.End);
```
[the remark this answers is not in the document]
Comment (gilmishal) [reply src/Http/WebUtilities/src/MultipartPipeReader.cs:124]: the buffer is being sliced within TryParseHeadersToEnd

File: src/Http/WebUtilities/src/MultipartPipeReader.cs [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+                }
+
+                if (line.IsSingleSegment)
+                {
+                    ParseHeaderLineFast(line.FirstSpan, ref accumulator);
+                    continue;
+                }
+
+                var lineReader = new SequenceReader<byte>(line);
+                ReadOnlySequence<byte> value;
```
Comment (jkotalik) [inline src/Http/WebUtilities/src/MultipartPipeReader.cs:257]: No need to declare value early.

File: src/Http/WebUtilities/src/MultipartPipeReader.cs [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+
+                if (!lineReader.TryReadTo(out ReadOnlySequence<byte> key, ColonDelimiter))
+                {
+                    throw new InvalidDataException($"Invalid header line: {GetStringFromReadOnlySequence(line)}");
+                }
+                value = line.Slice(lineReader.Position);
+
```
Comment (jkotalik) [inline src/Http/WebUtilities/src/MultipartPipeReader.cs:264]:
```suggestion
```

File: src/Http/WebUtilities/src/MultipartPipeReader.cs [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+            var sequenceReader = new SequenceReader<byte>(buffer);
+
+            while (!sequenceReader.End)
+            {
+                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> line, CrlfDelimiter))
+                {
+                    // Don't buffer indefinitely
+                    if (headersLength + sequenceReader.Consumed > HeadersLengthLimit)
+                    {
+                        throw new InvalidDataException($"Line length limit {HeadersLengthLimit - headersLength} exceeded.");
```
Comment (jkotalik) [inline src/Http/WebUtilities/src/MultipartPipeReader.cs:232]:
```suggestion
                    throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");
```

File: src/Http/WebUtilities/src/MultipartPipeReader.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/Http/WebUtilities/src/MultipartPipeReader.cs:120]: I still think there could be issues here. I think the code path for TryParseHeaderstoEndFast is fine, but I have concerns about TryParseheadersToEndSlow. At the end, you are doing `buffer = default`, which I think needs to be `buffer = buffer.Slice(sequenceReader.Position);`

File: src/Http/WebUtilities/src/MultipartPipeReader.cs [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+                value = line.Slice(lineReader.Position);
+
+
+                var decodedKey = GetStringFromReadOnlySequence(key);
+                var decodedValue = GetStringFromReadOnlySequence(value);
+
+                AppendAndVerify(ref accumulator, decodedKey, decodedValue);
+            }
+
+            buffer = default;
```
Comment (jkotalik) [inline src/Http/WebUtilities/src/MultipartPipeReader.cs:272]: Does this code path only execute when the sequence is empty?

File: src/Http/WebUtilities/src/MultipartPipeReader.cs [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+        /// <summary>
+        /// Read the next Multipart Section
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        public async Task<MultipartPipeSection?> ReadNextSectionAsync(CancellationToken cancellationToken = default)
+        {
+            if (_currentSectionReader == null) // first call.
+            {
+                // This section will drain any preamble data and remove the first boundary marker.
+                _currentSectionReader = new MultipartSectionPipeReader(_pipeReader, _boundary) { LengthLimit = HeadersLengthLimit };
```
Comment (jkotalik) [inline src/Http/WebUtilities/src/MultipartPipeReader.cs:67]: Idea (for future), could we get away with not having the MultipartSectionPipeReader class at all, tracking state in the MultipartPipeReader?

**Review by jkotalik — COMMENTED** [review]
Some initial comments. Still trying to get my head around the MultipartSectionPipeReader.

File: src/Http/WebUtilities/src/MultipartPipeReader.cs [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+            }
+
+            await _currentSectionReader.DrainAsync(cancellationToken);
+
+            if (_currentSectionReader.FinalBoundaryFound)
+            {
+                return null;
```
[the remark this answers is not in the document]
Comment (gilmishal) [reply src/Http/WebUtilities/src/MultipartPipeReader.cs:74]: In the pipeReader implementation we are reading to the end of the pipeReader once we find the finalBoundary, so in practice the behavior is the same, and there is no need to drain the trailing data - basically the difference is that the pipe is just drained to end as soon as the final boundary is found , instead of waiting for a subsequent DrainAsync.

https://github.com/dotnet/aspnetcore/blob/b784d8ea2d34513dd12a1f533a17431679c22c63/src/Http/WebUtilities/src/MultipartSectionPipeReader.cs#L187-L195

File: src/Http/WebUtilities/src/MultipartPipeReader.cs [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+                }
+
+                if (line.Length == 0) // an empty line means it's the end of the headers
+                {
+                    return true;
+                }
+
+                if (line.Length > lengthLimit)
+                {
+                    throw new InvalidDataException($"Line length limit {lengthLimit} exceeded.");
```
[the remark this answers is not in the document]
Comment (gilmishal) [reply src/Http/WebUtilities/src/MultipartPipeReader.cs:187]: I suspect these exceptions are the way they are since I tried sticking to the original behavior, which seems like a limitation from the way it was implemented with the Buffered Read Stream -

I don't think there is in fact a length limit for a specific header, and think it would make sense to use the same error message in all these situations and just use `throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");`

[the remark this answers is not in the document]
Comment (gilmishal) [reply src/Http/WebUtilities/src/MultipartPipeReader.cs:187]: And obviously update the tests I was reluctant to update

File: src/Http/WebUtilities/src/MultipartPipeReader.cs (hunk 14 above)
Comment (gilmishal) [reply src/Http/WebUtilities/src/MultipartPipeReader.cs:272]: It executes when the buffer is empty, or if we read till the end of the buffer, but didn't find all the headers - and didn't exceed the Headers' length limit.

File: src/Http/WebUtilities/src/MultipartPipeReader.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (gilmishal) [reply src/Http/WebUtilities/src/MultipartPipeReader.cs:120]: I think you are right. This certainly seems like a possible issue, I will also add a test for coverage.

[the remark this answers is not in the document]
Comment (gilmishal) [reply src/Http/WebUtilities/src/MultipartPipeReader.cs:120]: Definitely an issue, I managed to reproduce it when removing the ParseHeadersFast method and using a SingleByteStream - there were duplicate header values - updated the test as much as I could.
I do suggest adding tests forcing ParseHeadersSlow - but I am not sure how to do it, I don't know how to force multi segment - and I wouldn't want to change functionalities just for the test. Anyways the issue itself is fixed by your suggestion.

File: src/Http/WebUtilities/src/MultipartPipeReader.cs [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+            consumed = 0;
+
+            while (span.Length > 0)
+            {
+                // Find the end of the header.
+                var newLine = span.IndexOf(CrlfDelimiter);
+                ReadOnlySpan<byte> line;
+                var foundNewLine = newLine != -1;
+
+                if (foundNewLine)
```
[the remark this answers is not in the document]
Comment (gilmishal) [reply src/Http/WebUtilities/src/MultipartPipeReader.cs:162]: done

File: src/Http/WebUtilities/src/MultipartPipeReader.cs [hunk 19]
```diff
@@ -0,0 +1,329 @@
...
+                    var finishedParsing = TryParseHeadersToEnd(ref buffer, ref headersAccumulator, ref headersLength);
+
+                    if (finishedParsing)
+                    {
+                        return headersAccumulator.GetResults();
+                    }
+
+                    if (readResult.IsCompleted)
+                    {
+                        throw new InvalidDataException("Unexpected end of Stream, the content may have already been read by another component. ");
```
[the remark this answers is not in the document]
Comment (gilmishal) [reply src/Http/WebUtilities/src/MultipartPipeReader.cs]: done

File: src/Http/WebUtilities/src/MultipartPipeReader.cs (hunk 12 above)
Comment (gilmishal) [reply src/Http/WebUtilities/src/MultipartPipeReader.cs]: done

File: src/Http/WebUtilities/src/MultipartPipeReader.cs (hunk 11 above)
Comment (gilmishal) [reply src/Http/WebUtilities/src/MultipartPipeReader.cs]: done

**Comment by gilmishal** [comment]
@jkotalik any more comments? Can I continue to the next steps?

**Comment by jkotalik** [comment]
@gilmishal apologies, I'll get back to this within the week.

File: src/Http/WebUtilities/src/MultipartPipeSection.cs [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    /// <summary>
+    /// a Pipe Implementation of a Http Multipart Section
```
Comment (jkotalik) [inline src/Http/WebUtilities/src/MultipartPipeSection.cs:20]:
```suggestion
    /// A Pipe Implementation of a Http Multipart Section
```

File: src/Http/WebUtilities/src/MultipartSectionPipeReader.cs [hunk 21]
```diff
@@ -0,0 +1,443 @@
...
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    internal class MultipartSectionPipeReader : PipeReader
+    {
+        private static ReadOnlySpan<byte> CrlfDelimiter => new byte[] { (byte)'\r', (byte)'\n' };
```
Comment (jkotalik) [inline src/Http/WebUtilities/src/MultipartSectionPipeReader.cs:15]: See https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs for reference to another custom PipeReader implemented (doesn't directly implement PipeReader, but has the same methods).

**Review by jkotalik — COMMENTED** [review]
It's hard to follow what's going on in ReadAsync/TryRead/AdvanceTo. One thing I found is writing an example in code with a boundary and end-of-section to show how each of the cases are covered helps with understanding.

**Comment by jkotalik** [comment]
@gilmishal you get the chance to respond to this feedback?

**Comment by gilmishal** [comment]
Sorry, I'll respond tomorrow, been a crazy month.

**Comment by gilmishal** [comment]
> @gilmishal you get the chance to respond to this feedback?

Been a crazy month. I'll respond tomorrow.

**Comment by jkotalik** [comment]
I'm going to close this for now, feel free to reopen when you have the chance to review and respond to feedback.
